Schema Best Practices (#1546)
nasa-gcn/gcn.nasa.gov: app/routes/docs.notices.schema.practices.mdx

---
handle:
  breadcrumb: Schema Practices
---

# Best Practices for Schema

### Recommended Unit System for Your Schema

To maintain a consistent database for the same property reported by different instruments, it is best to primarily use International System of Units (SI) or standard units to ensure uniformity across all notices types. One can easily access unit conversion software, such as astropy, for internal team use.

### Choosing between "allOf", "anyOf", or "oneOf"

> We recommend using the "allOf" option because none of the core schema properties are specified as required. However, the flexibility of not defining the required properties while using "anyOf" can lead to false positives, as technically, anything will be validated against it.

Reviewer: I find this section a little cryptic. @dakota002, would you please help here?

> We recommend using the "allOf" option because none of the core schema properties are specified as required. However, the flexibility of not defining the required properties while using "anyOf" can lead to false positives, as technically, anything will be validated against it.

Reviewer 2: ```suggestion
JSON Schema supports the use of the keywords `allOf`, `anyOf`, and `oneOf` to combine schemas. These behave as logical AND, OR, and XOR operators. More often than not, if you are considering using existing Core schema elements, you will likely want to use the `allOf` option, as this will enforce that each listed sub-schema element is valid in your schema's structure. In our [schema definintion docs](/docs/notices/schema#define-your-schema) the `sample.schema.json` uses `allOf` to combine the Alert and Localization schemas with a couple additional properties.

It is important to keep in mind that including existing schema elements in your schema design will not override the requirement rules defined on those elements. If a property defined on an included sub-schema item is required in its own definition, it will also be required in your implementations.
```

Reviewer 2: @lpsinger, @Vidushi-GitHub , @jracusin Is this wording clear?

### Referencing Coincident Detection with a Different Instrument

We have designed a [follow-up core schema](https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/FollowUp.schema.json) for referencing instrument properties. Other instruments involved in follow-up studies can use this schema.

### Maintaining the Schema Version

> The GCN Team manages schema versioning. To stay informed about new releases, follow the [GCN-Schema](https://github.com/nasa-gcn/gcn-schema) repository for notifications. Updates are categorized into three levels: major, minor, and patch, depending on the kind of update according to [Semantic Versioning](https://semver.org/) rules.

Reviewer: @dakota002, @Courey, would you please help with making this section a bit more specific? I.e., explain that patch changes are always backwards and forward compatible, minor changes are always forwards compatible, and major changes break compatibility?

Also we need to communicate to producers that in their alerts they must refer to tagged, versioned, schema, and should _never_ refer to the main or latest stable version --- and explain why.

### Enforcing strict typing using unevaluatedProperties

> JSON Schema is typically flexible, and allow validation of JSON objects with undefined properties. In some cases this is fine, it can lead to potential false positives while running the validation step. To prevent this, we recommend adding the `unevaluatedProperties` key to your schema definition and setting its value to false. TThis key exists on the root level of the schema definition. Additionally, include `'$schema':true` in your schema's properties. You can find an example of this implementation in the `guano.schema.json` code for Swift/bat guano alerts in the [schema documentation](/docs/notices/schema).

Reviewer: There's a typo here: "TThis"

Also I find this section kind of cryptic.

---
handle:
  breadcrumb: Schema Practices
---

# Best Practices for Schema

### Recommended Unit System for Your Schema

To maintain a consistent database for the same property reported by different instruments, it is best to primarily use International System of Units (SI) or standard units to ensure uniformity across all notices types. One can easily access unit conversion software, such as astropy, for internal team use.

### Choosing between "allOf", "anyOf", or "oneOf"

We recommend using the "allOf" option because none of the core schema properties are specified as required. However, the flexibility of not defining the required properties while using "anyOf" can lead to false positives, as technically, anything will be validated against it.

### Referencing Coincident Detection with a Different Instrument

We have designed a [follow-up core schema](https://github.com/nasa-gcn/gcn-schema/blob/main/gcn/notices/core/FollowUp.schema.json) for referencing instrument properties. Other instruments involved in follow-up studies can use this schema.

### Maintaining the Schema Version

> The GCN Team manages schema versioning. To stay informed about new releases, follow the [GCN-Schema](https://github.com/nasa-gcn/gcn-schema) repository for notifications. Updates are categorized into three levels: major, minor, and patch, based on the nature of the changes. For example, major updates relate to core schema changes, minor updates involve new missions, and specific typing changes may be labeled as patches.

Author: ```suggestion
The GCN Team manages schema versioning. To stay informed about new releases, follow the [GCN-Schema](https://github.com/nasa-gcn/gcn-schema) repository for notifications. Updates are categorized into three levels: major, minor, and patch, depending on the kind of update according to [Semantic Versioning](https://semver.org/) rules. 
```

---
handle:
  breadcrumb: Schema Practices
---

> import { Highlight } from '~/components/Highlight'

Author: ```suggestion
```
